sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -115,6 +114,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
       Batch("Infer Filters", Once,
         InferFiltersFromConstraints) ::
       Batch("Operator Optimization after Inferring Filters", fixedPoint,
+        Seq(SimplifyBinaryComparison) ++
```

Comment:
We should apply `InferFiltersFromConstraints` first and then apply `SimplifyBinaryComparison` to retain `isnotnull` filter.

Reply:
I think the rules in `operatorOptimizationRuleSet`  are run after inferring filters and `SimplifyBinaryComparison` is in the rules, no?

Reply:
Because operatorOptimizationRuleSet will run double time. Once before and once after `InferFiltersFromConstraints`.

We should only do thie optimization after `InferFiltersFromConstraints`.